src/ray/gcs/pubsub/gcs_pub_sub.cc

```diff
@@ -0,0 +1,129 @@
+  rpc::GcsMessage message;
+  message.set_id(id.Binary());
+  std::string data_str;
+  data.SerializeToString(&data_str);
+  message.set_data(data_str);
+
+  auto on_done = [done](std::shared_ptr<CallbackReply> reply) {
+    if (done) {
+      done(Status::OK());
+    }
+  };
+
+  return redis_client_->GetPrimaryContext()->PublishAsync(
+      GenChannelPattern(channel, id), message.SerializeAsString(), on_done);
+}
+
+template <typename ID, typename Data>
+Status GcsPubSub<ID, Data>::Subscribe(const std::string &channel, const ID &id,
+                                      const Callback &subscribe,
+                                      const StatusCallback &done) {
+  return Subscribe(channel, boost::optional<ID>(id), subscribe, done);
+}
+
+template <typename ID, typename Data>
+Status GcsPubSub<ID, Data>::SubscribeAll(const std::string &channel,
+                                         const Callback &subscribe,
+                                         const StatusCallback &done) {
+  return Subscribe(channel, boost::none, subscribe, done);
+}
+
+template <typename ID, typename Data>
+Status GcsPubSub<ID, Data>::Unsubscribe(const std::string &channel, const ID &id,
+                                        const StatusCallback &done) {
+  if (done) {
+    unsubscribe_callbacks_[GenChannelPattern(channel, id)] = done;
+  }
+  return redis_client_->GetPrimaryContext()->PUnsubscribeAsync(
+      GenChannelPattern(channel, id));
+}
+
+template <typename ID, typename Data>
+Status GcsPubSub<ID, Data>::Subscribe(const std::string &channel,
+                                      const boost::optional<ID> &id,
+                                      const Callback &subscribe,
+                                      const StatusCallback &done) {
+  std::string pattern = GenChannelPattern(channel, id);
+  auto callback = [this, pattern, subscribe](std::shared_ptr<CallbackReply> reply) {
+    if (!reply->IsNil()) {
+      if (reply->GetMessageType() == "punsubscribe") {
+        if (unsubscribe_callbacks_.count(pattern)) {
+          unsubscribe_callbacks_[pattern](Status::OK());
+          unsubscribe_callbacks_.erase(pattern);
+        }
+        ray::gcs::RedisCallbackManager::instance().remove(
+            subscribe_callback_index_[pattern]);
+      } else {
+        const auto reply_data = reply->ReadAsPubsubData();
+        if (!reply_data.empty()) {
+          rpc::GcsMessage message;
+          message.ParseFromString(reply_data);
+          Data data;
+          data.ParseFromString(message.data());
+          subscribe(ID::FromBinary(message.id()), data);
+        }
+      }
+    }
+  };
+
+  int64_t out_callback_index;
+  auto status = redis_client_->GetPrimaryContext()->PSubscribeAsync(pattern, callback,
+                                                                    &out_callback_index);
+  if (id) {
+    subscribe_callback_index_[pattern] = out_callback_index;
```

Comment:
should make this class thread-safe, and document that in class comment.

Reply:
fixed